keyvalue/clients/src/main/java/tachyon/client/keyvalue/hadoop/KeyValueRecordWriter.java

```diff
@@ -23,34 +23,56 @@
 import org.apache.hadoop.util.Progressable;
 import org.apache.http.annotation.ThreadSafe;
 
+import tachyon.TachyonURI;
 import tachyon.client.keyvalue.KeyValueStoreWriter;
 import tachyon.client.keyvalue.KeyValueStores;
 import tachyon.exception.TachyonException;
 
 /**
- * A {@link RecordWriter} to write key-value pairs output by Reducers to a {@link KeyValueStores}.
+ * It writes key-value pairs into a temporary key-value store.
  */
 @ThreadSafe
 class KeyValueRecordWriter implements RecordWriter<BytesWritable, BytesWritable> {
-  private KeyValueStoreWriter mWriter;
-  private Progressable mProgress;
+  private static final KeyValueStores KEY_VALUE_STORES = KeyValueStores.Factory.create();
+
+  private final KeyValueStoreWriter mWriter;
+  private final Progressable mProgress;
 
   /**
    * Constructs a new {@link KeyValueRecordWriter}.
    *
-   * @param writer the key-value store writer
+   * @param storeUri the URI for the temporary key-value store to be created by this record writer
    * @param progress the object to be used for reporting progress
+   * @throws IOException when instance creation fails
    */
-  public KeyValueRecordWriter(KeyValueStoreWriter writer, Progressable progress) {
-    mWriter = writer;
+  public KeyValueRecordWriter(TachyonURI storeUri, Progressable progress) throws IOException {
+    try {
+      mWriter = KEY_VALUE_STORES.create(storeUri);
+    } catch (TachyonException e) {
+      throw new IOException(e);
+    }
     mProgress = progress;
   }
 
+  /**
+   * Copies a byte array of the specified length from the specified array, starting at offset 0.
+   *
+   * @param src the source array
+   * @param length the length of bytes to be copied
+   * @return the copied array
+   */
+  private byte[] copyBytes(byte[] src, int length) {
+    byte[] result = new byte[length];
+    System.arraycopy(src, 0, result, 0, length);
+    return result;
+  }
+
   @Override
   public synchronized void write(BytesWritable key, BytesWritable value) throws IOException {
     try {
-      mWriter.put(key.getBytes(), value.getBytes());
-      // Send a progress to the job manager to inform it that the task is still running.
+      mWriter.put(copyBytes(key.getBytes(), key.getLength()), copyBytes(value.getBytes(),
```

Comment:
is `key.getBytes().length == key.getLength()`? If so we could use `Arrays.clone(key.getBytes())` instead. If they are different, we can use `Arrays.copyOfRange`